File: js/src/maplestory/api/dto/character/characterCashItemEquipmentPresetDto.ts
import { CharacterCashItemEquipmentColoringPrismDto } from './characterCashItemEquipmentColoringPrismDto';
import { CharacterCashItemEquipmentOptionDto } from './characterCashItemEquipmentOptionDto';
import { CharacterCashItemEquipmentPresetDtoBody } from '../../response/character/characterCashItemEquipmentDtoBody';

/**
 * 캐릭터 프리셋 장착 캐시 장비 정보
 */
class CharacterCashItemEquipmentPresetDto {
  /**
   * 캐시 장비 부위 명
   */
  cashItemEquipmentPart: string;

  /**
   * 캐시 장비 슬롯 위치
   */
  cashItemEquipmentSlot: string;

  /**
   * 캐시 장비 명
   */
  cashItemName: string;

  /**
   * 캐시 장비 아이콘
   */
  cashItemIcon: string;

  /**
   * 캐시 장비 설명
   */
  cashItemDescription: string | null;

  /**
   * 캐시 장비 옵션
   */
  cashItemOption: CharacterCashItemEquipmentOptionDto[];

  /**
   * 캐시 장비 유효 기간 (KST)
   */
  dateExpire: string | null;

  /**
   * 캐시 장비 옵션 유효 기간 (KST, 시간 단위 데이터로 분은 일괄 0으로 표기)
   */
  dateOptionExpire: string | null;

  /**
   * 캐시 장비 라벨 정보
   */
  cashItemLabel: string | null;

  /**
   * 캐시 장비 컬러링프리즘 정보
   */
  cashItemColoringPrism: CharacterCashItemEquipmentColoringPrismDto | null;

  /**
   * 아이템 장착 가능 성별
   */
  itemGender: string | null;

  /**
   * 스킬명
   */
  skills: string[];

  constructor(obj: CharacterCashItemEquipmentPresetDtoBody) {
    const {
      cash_item_equipment_part,
      cash_item_equipment_slot,
      cash_item_name,
      cash_item_icon,
      cash_item_description,
      cash_item_option,
      date_expire,
      date_option_expire,
      cash_item_label,
      cash_item_coloring_prism,
      item_gender,
      skills,
    } = obj;

    this.cashItemEquipmentPart = cash_item_equipment_part;
    this.cashItemEquipmentSlot = cash_item_equipment_slot;
    this.cashItemName = cash_item_name;
    this.cashItemIcon = cash_item_icon;
    this.cashItemDescription = cash_item_description;
    this.cashItemOption = cash_item_option.map(
      (option) => new CharacterCashItemEquipmentOptionDto(option),
    );
    this.dateExpire = date_expire;
    this.dateOptionExpire = date_option_expire;
    this.cashItemLabel = cash_item_label;
    this.cashItemColoringPrism = cash_item_coloring_prism
      ? new CharacterCashItemEquipmentColoringPrismDto(cash_item_coloring_prism)
      : null;
    this.itemGender = item_gender;
    this.skills = skills;
  }
}

export { CharacterCashItemEquipmentPresetDto };

File: java/src/main/java/dev/spiralmoon/maplestory/api/kms/dto/character/CharacterCashItemEquipmentPresetDTO.java
package dev.spiralmoon.maplestory.api.kms.dto.character;

import com.fasterxml.jackson.annotation.JsonProperty;
import com.fasterxml.jackson.annotation.JsonSetter;
import com.fasterxml.jackson.annotation.Nulls;
import dev.spiralmoon.maplestory.api.Utils;
import lombok.AllArgsConstructor;
import lombok.Data;
import lombok.NoArgsConstructor;
import lombok.ToString;

import java.time.ZonedDateTime;
import java.util.List;

/**
 * 캐릭터 프리셋 장착 캐시 장비 정보
 */
@AllArgsConstructor
@NoArgsConstructor
@Data
@ToString
public class CharacterCashItemEquipmentPresetDTO implements dev.spiralmoon.maplestory.api.common.dto.character.CharacterCashItemEquipmentPresetDTO<
        CharacterCashItemEquipmentOptionDTO,
        CharacterCashItemEquipmentColoringPrismDTO> {

    /**
     * 캐시 장비 부위 명
     */
    @JsonProperty("cash_item_equipment_part")
    private String cashItemEquipmentPart;

    /**
     * 캐시 장비 슬롯 위치
     */
    @JsonProperty("cash_item_equipment_slot")
    private String cashItemEquipmentSlot;

    /**
     * 캐시 장비 명
     */
    @JsonProperty("cash_item_name")
    private String cashItemName;

    /**
     * 캐시 장비 아이콘
     */
    @JsonProperty("cash_item_icon")
    private String cashItemIcon;

    /**
     * 캐시 장비 설명
     */
    @JsonProperty("cash_item_description")
    private String cashItemDescription;

    /**
     * 캐시 장비 옵션 목록
     */
    @JsonProperty("cash_item_option")
    @JsonSetter(nulls = Nulls.AS_EMPTY)
    private List<CharacterCashItemEquipmentOptionDTO> cashItemOption;

    /**
     * 캐시 장비 유효 기간 (KST)
     */
    @JsonProperty("date_expire")
    private String dateExpire;

    /**
     * 캐시 장비 옵션 유효 기간 (KST, 시간 단위 데이터로 분은 일괄 0으로 표기)
     */
    @JsonProperty("date_option_expire")
    private String dateOptionExpire;

    /**
     * 캐시 장비 라벨 정보
     */
    @JsonProperty("cash_item_label")
    private String cashItemLabel;

    /**
     * 캐시 장비 컬러링프리즘 정보
     */
    @JsonProperty("cash_item_coloring_prism")
    private CharacterCashItemEquipmentColoringPrismDTO cashItemColoringPrism;

    /**
     * 캐시 장비(무기) 이펙트 프리즘 정보
     */
    @JsonProperty("cash_item_effect_prism")
    private CharacterCashItemEquipmentEffectPrismDTO cashItemEffectPrism;

    /**
     * 아이템 장착 가능 성별
     */
    @JsonProperty("item_gender")
    private String itemGender;

    /**
     * 스킬명
     */
    @JsonProperty("skills")
    @JsonSetter(nulls = Nulls.AS_EMPTY)
    private List<String> skills;

    /**
     * 프리스타일 쿠폰 적용 여부 (0:미적용, 1:적용)
     */
    @JsonProperty("freestyle_flag")
    private String freestyleFlag;

    /**
     * 표정 합성 정보
     */
    @JsonProperty("emotion_name")
    private String emotionName;

    /**
     * 캐시 장비 유효 기간 (KST)
     */
    public ZonedDateTime getDateExpire() {
        if (this.dateExpire != null && !"expired".equals(this.dateExpire)) {
            return Utils.toZonedDateTime(this.dateExpire);
        } else {
            return null;
        }
    }

    /**
     * 캐시 장비 유효 기간 만료 여부
     */
    public Boolean isExpired() {

        if (this.dateExpire == null) {
            return null;
        }

        return "expired".equals(this.dateExpire);
    }

    /**
     * 캐시 장비 옵션 유효 기간 (KST, 시간 단위 데이터로 분은 일괄 0으로 표기)
     */
    public ZonedDateTime getDateOptionExpire() {
        if (this.dateOptionExpire != null && !"expired".equals(this.dateOptionExpire)) {
            return Utils.toZonedDateTime(this.dateOptionExpire);
        } else {
            return null;
        }
    }

    /**
     * 캐시 장비 옵션 유효 기간 만료 여부
     */
    public Boolean isOptionExpired() {

        if (this.dateOptionExpire == null) {
            return null;
        }

        return "expired".equals(this.dateOptionExpire);
    }

    /**
     * 프리스타일 쿠폰 적용 여부
     */
    public boolean isFreestyleFlag() {
        return this.freestyleFlag.equals("1");
    }
}
